[ResponseOps] ES|QL rule type improvements - write query results to the alert doc

doakalexi opened the pull request:
Resolves https://github.com/elastic/response-ops-team/issues/200

## Summary

This PR copies the fields from the ES|QL query results to the alert doc. Now that we are writing everything to the alert doc I removed the source fields selector from the ES|QL ui.


### Checklist

- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### To verify

- Create an ES Query ES|QL rule.
- Go to [dev tools](http://localhost:5601/app/dev_tools#/console) and run the query below to verify that the query results are written to the alert doc.
```
GET .internal.alerts-stack.alerts-default*/_search
```

doakalexi commented:
/ci

doakalexi commented:
/ci

doakalexi commented:
/ci

pmuellr commented on x-pack/test/alerting_api_integration/spaces_only/tests/alerting/group3/builtin_alert_types/es_query/esql_only.ts:
```diff
@@ -260,6 +250,7 @@ export default function ruleTests({ getService }: FtrProviderContext) {
       expect(alertDoc['host.name']).to.eql(['host-1']);
       expect(alertDoc['host.hostname']).to.eql(['host-1']);
       expect(alertDoc['host.id']).to.eql(['1']);
+      expect(alertDoc.c[0]).greaterThan(0);
```
Does this mean a query can overwrite ANY field in the alertDoc?  And I guess add new fields not defined in mappings?  Neither of those seem right ...

doakalexi replied:
I wanted to acknowledge for the PR that we are talking offline about this 🙂

umbopepato commented:
@doakalexi I'm getting this error every time the test rule tries to save alerts:

<img width="1485" alt="image" src="https://github.com/elastic/kibana/assets/18363145/3dba35ab-c196-4118-b3c7-973794bd4e64">

These are the rule params I'm using:

<img width="612" alt="image" src="https://github.com/elastic/kibana/assets/18363145/48736412-6e44-4453-9883-01d9cb8bd59e">

pmuellr commented:
Umbo points out ^^^ another thing we'll have to check.  Is it an ECS field, that we allow you to update, and **is it the right type**?  I'd think we'd want to exclude fields that are the wrong type.  For example, if a number is expected, but the value being added is a string (that doesn't parse into a valid number - oh man, this is getting complicated!), then don't add that field to the alert.

mikecote commented:
>Umbo points out ^^^ another thing we'll have to check. Is it an ECS field, that we allow you to update, and is it the right type? I'd think we'd want to exclude fields that are the wrong type. For example, if a number is expected, but the value being added is a string (that doesn't parse into a valid number - oh man, this is getting complicated!), then don't add that field to the alert.

Does the ignore_malformed help us in this scenario? https://github.com/elastic/kibana/issues/161465

doakalexi commented:
Thanks for the feedback, I am sorry I went down a different path than we are expecting. I think things should be working as expected now.

I updated the changes in this [commit]([8af0458](https://github.com/elastic/kibana/pull/184541/commits/8af0458b0c8b28bbe45e34d62fca1239054a48a6)) to only write fields that we have mapped. We intersect result fields with mapped fields and copy array of values into alert field.

Currently, we do not map all ECS fields for ES Query alerts so I added the `useEcs: true` to the stack alerts AAD config.

I confirmed with @shanisagiv1 that we want to remove the ui code to select source fields for ES|QL.

pmuellr commented:
> Does the ignore_malformed help us in this scenario? https://github.com/elastic/kibana/issues/161465

Well, sure, but that will be likely very difficult to diagnose if we have weird problems where sometimes fields are accepted, and sometimes aren't.

I'd say if we need to do that for the field TYPE checking - providing non-numeric-parseable strings as numeric types - then ... fine.  :-)  I can imagine that it might be hard-to-impossible for us to do real type checking here.  

But we should certainly be filtering out invalid / unwanted field names without depending on ignore_malformed ...

doakalexi commented on x-pack/test/api_integration/apis/maps/maps_telemetry.ts:
```diff
@@ -36,8 +36,8 @@ export default function ({ getService }: FtrProviderContext) {
       );
-      expect(geoPointFieldStats.count).to.be(39);
-      expect(geoPointFieldStats.index_count).to.be(10);
+      expect(geoPointFieldStats.count).to.be(47);
```
I think they changed bc we are now mapping the ecs fields for the stack alert aad index

umbopepato reviewed (approved):
Works as expected, LGTM!

pmuellr commented:
I haven't finished reviewing the new code (since last review), will do that next week.

I did try it out though.  I used [`es-apm-sys-sim 1 4 -k`](https://github.com/pmuellr/es-apm-sys-sim) to start generating some documents in the index `es-apm-sys-sim`.  This will generate docs with 4 (from the 2nd param) `host.name` fields like `host-1`, `host-2`, and some other simulated metric data.

Then I built an ES|QL rule with query:

```
from es-apm-sys-sim |
rename system.cpu.total.norm.pct as host.cpu.usage |
limit 3
```

(the rename changes the tool-generated index's value to an ECS-named value)

When I look at the alert document, I see the following fields that got added per the PR:

```
host.name: ["host-4","host-3","host-1"],
host.cpu.usage: [ 0.5, 1 ]
```

The `host.name` is easy to understand.  It picked 3 documents (of 1000's) that happened to have different `host.name` values.  Lucky.  But it did the "array-ization" correctly! (of course!) \o/

The `host.cpu.usage` is curious though.  Turns out, host 1 had cpu of 1, and hosts 3 and 4 had 0.5.  So it "flattened" the `host.cpu.usage` values.  Seems like this must have been from our code, but not sure yet.

Even if we changed this to produce what I would expect of `[0.5, 0.5, 1]` (the missing `0.5` value), the results are still of questionable value.  I'm sure you can build some useful alerts with it in this shape, but I'm guessing people will want to see the tie between the host name and cpu usage, which you can't today.

So - maybe a bug in the code where it didn't produce 3 values for cpu usage.  And a longer term question of how useful it will be to do this simple aggregation of matching field values.

doakalexi commented:
> So - maybe a bug in the code where it didn't produce 3 values for cpu usage. And a longer term question of how useful it will be to do this simple aggregation of matching field values.

Based on the issue, I think flattening the values is how it's supposed to work. @mikecote or @shanisagiv1 maybe can confirm what behavior we want here

pmuellr commented on x-pack/plugins/stack_alerts/public/rule_types/es_query/expression/esql_query_expression.tsx:
```diff
@@ -61,12 +58,11 @@ export const EsqlQueryExpression: React.FC<
     termSize: DEFAULT_VALUES.TERM_SIZE,
     searchType: SearchType.esqlQuery,
-    sourceFields: sourceFields ?? DEFAULT_VALUES.SOURCE_FIELDS,
+    sourceFields: DEFAULT_VALUES.SOURCE_FIELDS,
```
`DEFAULT_VALUES.SOURCE_FIELDS` seems to be `[]`, and this seems to be the only reference to it.

doakalexi replied:
Oh true, I can remove it.

doakalexi replied:
Resolved in this commit https://github.com/elastic/kibana/pull/184541/commits/e84f29566ccbd83b7d4247303006192474008274

pmuellr reviewed (approved):
code LGTM, but have some discussion topics in the PR thread ...

pmuellr commented:
I tried this out as follows.  I used [`es-apm-sys-sim 1 4 -k`](https://github.com/pmuellr/es-apm-sys-sim) to start generating some documents in the index `es-apm-sys-sim`.  This will generate docs with 4 (from the 2nd param) `host.name` fields like `host-1`, `host-2`, and some other simulated metric data.

Then I built an ES|QL rule with query:

```
from es-apm-sys-sim |
rename system.cpu.total.norm.pct as host.cpu.usage |
limit 3
```

(the rename changes the tool-generated field name that has values, to an ECS name)

When I look at the alert document, I see the following fields that got added per the PR:

```
host.name: ["host-4","host-3","host-1"],
host.cpu.usage: [ 0.5, 1 ]
```

The `host.name` is easy to understand.  It picked 3 documents (of 1000's) that happened to have different `host.name` values.  Lucky.  But it did the "array-ization" correctly.

The `host.cpu.usage` is curious though.  Turns out, host 1 had cpu of 1, and hosts 3 and 4 had 0.5.  So it "flattened" the `host.cpu.usage` values.  Seems like this must have been from our code, but not sure yet.

Even if we changed this to produce what I would expect of `[0.5, 0.5, 1]` (the missing `0.5` value), the results are still of questionable usage.  I'm sure you can build some useful alerts with it in this shape, but I'm guessing people will want to see the tie between the host name and cpu usage, which you can't today.

So - maybe a bug in the code where it didn't produce 3 values for cpu usage?  I'm not sure this is a big deal, since having the somewhat "mess" of ECS values isn't going to be precise enough anyway - does it really matter if they are flattened as well as not really organized?  I also couldn't find (quickly) where this might be done in our code, but assume it's our code and not an artifact of ES|QL ...

And a longer term question of figuring how to get these values organized better ... thinking we should solve this when we add "grouping", where we have similar issues to tackle.

doakalexi commented on x-pack/plugins/stack_alerts/server/rule_types/es_query/lib/fetch_esql_query.ts:
```diff
@@ -60,10 +64,10 @@ export async function fetchEsqlQuery({
+        hits,
       },
       resultLimit: alertLimit,
-      sourceFieldsParams: params.sourceFields,
```
It is still used by the other es query types, just removed for ES|QL. I will add a comment to say the param is ignored now for ES|QL

doakalexi replied:
Resolved in this commit

doakalexi commented:
> Even if we changed this to produce what I would expect of `[0.5, 0.5, 1]` (the missing `0.5` value), the results are still of questionable usage. I'm sure you can build some useful alerts with it in this shape, but I'm guessing people will want to see the tie between the host name and cpu usage, which you can't today.
> 
> So - maybe a bug in the code where it didn't produce 3 values for cpu usage? I'm not sure this is a big deal, since having the somewhat "mess" of ECS values isn't going to be precise enough anyway - does it really matter if they are flattened as well as not really organized? I also couldn't find (quickly) where this might be done in our code, but assume it's our code and not an artifact of ES|QL ...
>
> And a longer term question of figuring how to get these values organized better ... thinking we should solve this when we add "grouping", where we have similar issues to tackle.

Yeah I definitely see that it might not be as useful if you can't see the link between host.name and cpu usage. This code is in https://github.com/elastic/kibana/blob/main/x-pack/plugins/triggers_actions_ui/common/data/lib/parse_aggregation_results.ts#L88. If we don't want to flatten I can go ahead and change this. I am wondering if @mikecote or @shanisagiv1 have any opinions or ideas, bc the example in the issue is flattened.

mikecote commented:
What does the alternative look like if we don't flatten, do we create new fields?

doakalexi commented:
> What does the alternative look like if we don't flatten, do we create new fields?
Using patrick's example it would return [0.5, 0.5, 1] instead of [0.5, 1]

doakalexi commented:
hi @nreese can I get a review on the test that changed in this PR when you get a chance pls

pmuellr commented:
> Yeah I definitely see that it might not be as useful if you can't see the link between host.name and cpu usage

I think that's probably a decent reason to flatten.  Since we can't get the "paired" fields together, it might be nicer to get the flattened version, having it be less "noisy".  Once we figure out grouping, hopefully we'll have a little better story :-)

We should definitely doc this behavior - and I forgot to mention, we need to doc how this works in general, I think?  Not sure where we talk about the data in alerts.  Seems like we'd want to provide an example of converting a field to an alert-doc'able field.  Fine as a follow-on issue.

nreese reviewed (approved):
kibana-gis changes LGTM
code review only

doakalexi commented:
Hi y'all, I got confirmation from @shanisagiv1 that we don't want to flatten the data. I updated the code in this commit [4f46742](https://github.com/elastic/kibana/pull/184541/commits/4f467427a70fcee66649b36045872fdda6339eea). Thanks!
